### ISSUE TITLE
Can't use Traverson to navigate templated links produced by Spring Hateoas

### ISSUE THREAD

**Issue by cgfrost** [spring-projects/spring-hateoas#579]
I have an API being produced by Spring Hateoas with `HypermediaType.HAL` enabled and I'm trying to navigate the API using a client written using Traverson. Both the API and the client are using version 0.23.0.RELEASE running in Spring Boot Apps.

I'm following a hop and Traverson finds the link, for the given rel, which look like this `http://localhost:3110/accounts/{customer-uuid}/{account-uuid}` and gets to the single arg constructor in `UriTemplate` (line 54) where a matcher/regex is applied to find path variables but it doesn't find any. Further on in the execution it tries to expand the above Uri (line 193) and throws an IllegalArgumentException due to the first `{` char instead of substituting the parameters from the hop. 

I would expect to be able to round trip Spring Hateoas library, such that the client can navigate an generated API. I've read through the guide, reference manual and a fair bit of the JavaDoc and don't think I've missed anything but if I have please consider this a bug to improve the docs.

**Comment by gregturn** [comment]
Got a sample project we can use?

**Comment by cgfrost** [comment]
I thought you might ask that. Let me split out something small from the two applications and I'll provide them.

**Comment by gregturn** [comment]
;)

**Comment by gregturn** [comment]
I may be running into this right now. It appears Spring HATEOAS cannot read a templated link, but can generate them.

**Comment by cgfrost** [comment]
Yeah, that's what I've seen. `UriTemplate` fails to spot the templates in the href. I have some spare time on Thursday, was planning to split out an app to recreate it then....

**Comment by gregturn** [comment]
Not precisely. A structure like:

```
search : {
    href : "/orders/{q}"
    templated : true
}
```
...will trip the HAL deserializer, since it flat out isn't looking for link templates. The Affordances PR (https://github.com/spring-projects/spring-hateoas/tree/rebase/affordances) introduces another mediatype, HAL Forms, that DOES handle this. When trying to align that mediatype to leverage HAL's already built `List<Link>` ser/deser, a whole bunch of test cases broke.

Pulling the bits from that PR into HAL resulting in an old HAL test case breaking, which I discovered to have an incorrect HAL document.

Hence, I've patched HAL to handle this properly, and HAL Forms will inherit the benefit. It's probably going to take a couple more weeks to sift through this given this branch has LOT.

**Comment by gregturn** [comment]
Bottom line: not sure you need to spend the effort given I just reproduced your issue today. This presumes you are intentionally letting Traverson generate a templated link. All the test cases I built around that assumed that ALL link parameters are expanded.

**Comment by cgfrost** [comment]
At the high level I'm trying to follow a hop, as part of that I want the link to be expanded with the template parameters in the hop. I won't pull out some apps then but happy to test a snapshot at some point if that helps.

**Comment by gregturn** [comment]
Per the ref docs, you can supply parameters at each hop, or a universal map applied to ALL hops.

```
ParameterizedTypeReference<Resource<Item>> resourceParameterizedTypeReference = new ParameterizedTypeReference<Resource<Item>>() {};

Resource<Item> itemResource = traverson.//
		follow(rel("items").withParameter("projection", "noImages")).//
		follow("$._embedded.items[0]._links.self.href").//
		toObject(resourceParameterizedTypeReference);
```

Using the static **rel** function gives you access to the **withParameter** option (which is chainable).

The question to ask is, what does it mean to navigate a templated link without expanding it? To me, you either want to throw away the parameter, or you want to swap it out with something.

For example, if you were using Spring Data REST and had projections, `{?projection}` might appear in all your links. In that case, you might wish to expand it with nothing (throw it away) or expand it with a certain view applied to every query.

**Comment by cgfrost** [comment]
I think we have missunderstood. I'm not trying to navigate a templated link without expanding it. I'm using the approach you have described to give some parameters to the hop so the link can be expanded. This is failing because the link doesn't recognize itself as being templated and never tries to fill in the parameters.
